### ISSUE TITLE
@defer on two inline fragments on the query type swallows the second field

### ISSUE THREAD

**Issue by tobias-tengler** [ChilliCream/hotchocolate#4491]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I want to preface this by saying that I'm not even sure if this is a valid usage of @defer. Please let me know if it isn't!

I was just playing around whether #4485 does also apply to non-list items, when I ran into the following issue.

Sending a query with two deferred fields from within BCP correctly resolves the first field and then bails on the second with an "Multipart error".

![image](https://user-images.githubusercontent.com/45513122/143767554-4cc1aa48-e995-46f4-9295-fff13301849a.png)

![image](https://user-images.githubusercontent.com/45513122/143767532-78776bf9-70e1-4a7f-b3a5-0285b55c347a.png)

![image](https://user-images.githubusercontent.com/45513122/143767545-c3cd72d5-b459-4fab-a56a-0cb51a31230a.png)

Looking at the network log, I can only see the first patch being sent:

![image](https://user-images.githubusercontent.com/45513122/143767782-4f7a6d6f-9dcb-4735-aa55-9191a94ab35d.png)

Looking at the Hot Chocolate logs I can see the following exception being thrown (full details below): `System.ArgumentException: Invalid fragment id. (Parameter 'fragmentId')`

### Steps to reproduce

1.  Clone the following branch/repository: https://github.com/tobias-tengler/hc-defer-dataloader-issue
2. Start the project
3. Execute the following query

```graphql
{
  ...@defer {
    field1
  }
  ...@defer {
    field2
  }
}
```

### Relevant log output

```shell
An unhandled exception has occurred while executing the request.
      System.ArgumentException: Invalid fragment id. (Parameter 'fragmentId')
         at HotChocolate.Execution.Processing.Plan.QueryPlan.GetDeferredPlan(Int32 fragmentId)
         at HotChocolate.Execution.Processing.DeferredFragment.ExecuteAsync(IOperationContext operationContext)
         at HotChocolate.Execution.Processing.DeferredTaskExecutor.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
         at HotChocolate.Execution.Processing.DeferredTaskExecutor.GetAsyncEnumerator(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at HotChocolate.Execution.DeferredQueryResult.EnumerateResults.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
         at HotChocolate.Execution.DeferredQueryResult.EnumerateResults.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
         at HotChocolate.Execution.DeferredQueryResult.EnumerateResults.GetAsyncEnumerator(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at HotChocolate.Execution.Serialization.MultiPartResponseStreamSerializer.WriteResponseStreamAsync(IResponseStream responseStream, Stream outputStream, CancellationToken cancellationToken)
         at HotChocolate.Execution.Serialization.MultiPartResponseStreamSerializer.WriteResponseStreamAsync(IResponseStream responseStream, Stream outputStream, CancellationToken cancellationToken)
         at HotChocolate.AspNetCore.Serialization.DefaultHttpResultSerializer.SerializeAsync(IExecutionResult result, Stream stream, CancellationToken cancellationToken)
         at HotChocolate.AspNetCore.MiddlewareBase.WriteResultAsync(HttpResponse response, IExecutionResult result, Nullable`1 statusCode, CancellationToken cancellationToken)      
         at HotChocolate.AspNetCore.HttpPostMiddleware.HandleRequestAsync(HttpContext context, AllowedContentType contentType)
         at HotChocolate.AspNetCore.HttpPostMiddleware.HandleRequestAsync(HttpContext context, AllowedContentType contentType)
         at HotChocolate.AspNetCore.HttpPostMiddleware.InvokeAsync(HttpContext context)
         at HotChocolate.AspNetCore.WebSocketSubscriptionMiddleware.InvokeAsync(HttpContext context)
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

### Product

Hot Chocolate

### Version

12.3.2

**Comment by tobias-tengler** [comment]
I'm not sure what a fragmentId is, but I also tried with named fragments - same result:
```graphql
{
  ...Field1 @defer
  ...Field2 @defer
}

fragment Field1 on Query {
  field1
}

fragment Field2 on Query {
  field2
}

```

**Comment by tobias-tengler** [comment]
The issue only exists on the root type, the following works as expected:
```graphql
{
  complexType {
    ... @defer {
      complexField1
    }
    ... @defer {
      complexField2
    }
  }
}
```

I say it works as expected, but that's not entirely true. Both fields are resolved, BUT sequentially, which doesn't make much sense IMO.

**Comment by tobias-tengler** [comment]
I didn't want to raise an issue yet, because I'm not sure if the following is actually spec compliant behavior or not (I find it really hard to switch between spec PRs and find scattered information on this).

Is it an issue that the order of fields in the request changes in the final payload?

![image](https://user-images.githubusercontent.com/45513122/143768656-ed915757-c05e-48d3-888b-7a182f86ec5b.png)

Looking at the patches I don't see how you could order this correctly though:
```
---
Content-Type: application/json; charset=utf-8

{"data":{"complexType":{}},"hasNext":true}
---
Content-Type: application/json; charset=utf-8

{"data":{"field1":"field1"},"hasNext":true}
---
Content-Type: application/json; charset=utf-8

{"path":["complexType"],"data":{"complexField1":"complexfield1"},"hasNext":false}
-----
```

if this is already covered by the spec, please link me to it. I don't even remember if the original spec demands the order of fields to be the same between request and response atm 😅

**Comment by tobias-tengler** [comment]
These are still valid concerns or potential bugs with the `@defer` implementation, so I'm marking this issue as non-stale.

**Comment by michaelstaib** [comment]
This one is fixed. It was a bug in the operation compiler that is now fixed in the execution engine branch.
https://github.com/ChilliCream/hotchocolate/pull/5168

**Comment by tobias-tengler** [comment]
I can verify that the exception no longer happens in the latest version (13.0.0-preview.33). But the actual deferring seems still broken. If I `Task.Delay` the execution of both fields and make one slower than the other, they both pop in only after the slowest one has been resolved. 
![defer](https://user-images.githubusercontent.com/45513122/186874570-1a046ce9-29c3-4e45-ab52-712e39f1d48a.gif)
You can see the multipart chunks, but the response is not really streaming. Might be a transport bug.

**Comment by tobias-tengler** [comment]
@michaelstaib does the defer spec say anything about guaranteed ordering of fields? Because if both deferred fields have a similar timing, they fields might be swapped in the response:

![defer2](https://user-images.githubusercontent.com/45513122/186875289-3a50fb7a-8b68-4138-86a8-4b814cd637f3.gif)

**Comment by tobias-tengler** [comment]
Resolved in 13.0.0-preview.58!
